File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int[][] map;
    static boolean[][] visited;
    static int[] dirY = {-1, 0, +1, 0};
    static int[] dirX = {0, +1, 0, -1};
    static int n;
    static int l;
    static int r;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        l = Integer.parseInt(st.nextToken());
        r = Integer.parseInt(st.nextToken());

        map = new int[n][n];
        visited = new boolean[n][n];

        for(int y=0; y<n; y++){
            st = new StringTokenizer(br.readLine());
            for(int x=0; x<n; x++){
                map[y][x] = Integer.parseInt(st.nextToken());
            }
        }

        int answer = 0;
        boolean flag = true;
        while(flag) {
            flag = false;
            answer++;

            for(boolean[] x : visited){
                Arrays.fill(x, false);
            }

            for (int y = 0; y < n; y++) {
                for (int x = 0; x < n; x++) {
                    if (!visited[y][x]) {
                        if(bfs(y, x)) flag = true;
                    }
                }
            }
        }

        System.out.println(answer-1);
    }

    public static boolean bfs(int startY, int startX){
        ArrayList<int[]> country = new ArrayList<>();
        visited[startY][startX] = true;
        country.add(new int[]{startY, startX});

        int p = 0;
        int sum = 0;
        while(p < country.size()) {
            int[] now = country.get(p);
            sum += map[now[0]][now[1]];
            p++;

            for (int i = 0; i < 4; i++) {
                int newY = now[0] + dirY[i];
                int newX = now[1] + dirX[i];

                if (newY < 0 || newY >= n || newX < 0 || newX >= n) continue;
                int dif = Math.abs(map[newY][newX] - map[now[0]][now[1]]);

                if (!visited[newY][newX] && dif >= l && dif <= r) {
                    visited[newY][newX] = true;
                    country.add(new int[]{newY, newX});
                }
            }
        }

        int newSum = sum / country.size();
        for(int[] x: country){
            map[x[0]][x[1]] = newSum;
        }

        return p != 1;
    }
}

File: 백준/Gold/16234. 인구 이동/인구 이동.py
from collections import deque

N, L, R = map(int, input().split())

board = [list(map(int, input().split())) for _ in range(N)]
visited = [[0]*N for _ in range(N)]
sumArr = [0] * (N*N+1)

dy = [-1, 0, +1, 0]
dx = [0, +1, 0, -1]
sum = 0
cnt = 0

def bfs(y, x, seq):
    global sum, cnt
    flag = True

    que = deque()
    que.append([y, x, board[y][x]])
    visited[y][x] = seq
    sum += board[y][x]
    cnt += 1
    
    while que:
        nowY, nowX, nowV = que.popleft()
        for i in range(4):
            newY = nowY + dy[i]
            newX = nowX + dx[i]

            if newY < 0 or newY >= N or newX < 0 or newX >= N:
                continue
            newV = board[newY][newX]
            if visited[newY][newX] > 0 or abs(nowV - newV) < L or abs(nowV - newV) > R:
                continue

            flag = False
            sum += newV
            cnt += 1
            visited[newY][newX] = seq
            que.append([newY, newX, newV])
    
    return flag


def move():
    for y in range(N):
        for x in range(N):
            if visited[y][x] > 0:
                board[y][x] = sumArr[visited[y][x]]

answer = 0
while True:
    seq = 0
    visited = [[0] * N for _ in range(N)]
    flag = True
    sumArr = [0] * (N*N+1)

    for y in range(N):
        for x in range(N):
            if visited[y][x] > 0:
                continue

            sum = 0
            cnt = 0
            seq += 1
            if not bfs(y, x, seq):
                flag = False
            sumArr[seq] = sum // cnt

    move()

    if not flag:
        answer += 1
    else:
        break

print(answer)